src/replica.rs

```diff
@@ -372,75 +447,81 @@ impl<K: KVStore, A: App<K> + 'static, N: Network + 'static> ReplicaSpec<K, A, N>
             self.on_commit_block,
             self.on_prune_block,
             self.on_update_highest_qc,
-            self.on_update_locked_view,
+            self.on_update_locked_qc,
+            self.on_update_highest_tc,
             self.on_update_validator_set,
             self.on_propose,
             self.on_nudge,
             self.on_vote,
             self.on_new_view,
+            self.on_timeout_vote,
+            self.on_advance_view,
             self.on_receive_proposal,
             self.on_receive_nudge,
             self.on_receive_vote,
             self.on_receive_new_view,
+            self.on_receive_timeout_vote,
+            self.on_receive_advance_view,
             self.on_start_view,
             self.on_view_timeout,
             self.on_collect_qc,
+            self.on_collect_tc,
             self.on_start_sync,
             self.on_end_sync,
             self.on_receive_sync_request,
-            self.on_send_sync_response
+            self.on_send_sync_response,
         );
 
-        let (event_publisher, event_subscriber) = 
-            if !event_handlers.is_empty() {
-                Some(mpsc::channel()).unzip()
-            } else { (None, None) };
+        let (event_publisher, event_subscriber) = if !event_handlers.is_empty() {
+            Some(mpsc::channel()).unzip()
+        } else {
+            (None, None)
+        };
 
         let (block_sync_server_shutdown, block_sync_server_shutdown_receiver) = mpsc::channel();
-        let mut block_sync_server = BlockSyncServer::new(
-            block_sync_server_config, 
-            BlockTreeCamera::new(self.kv_store.clone()), 
-            block_sync_requests, 
+        let block_sync_server = BlockSyncServer::new(
+            block_sync_server_config,
+            BlockTreeCamera::new(self.kv_store.clone()),
+            block_sync_requests,
             self.network.clone(),
             block_sync_server_shutdown_receiver,
-            event_publisher.clone()
+            event_publisher.clone(),
         );
         let block_sync_server = block_sync_server.start();
```

Comment:
We have to repeat this sequence of "first, call `new`. Then, call `start`" for both `BlockSyncServer` and `Algorithm`. I think this is awkward. I think it'd be better if `Algorithm` and `BlockSyncServer` totally encapsulate how they use threading, instead of giving callers a `JoinHandle`. These structs should implement `Drop`.

We should also similarly fully encapsulate the poller thread and the event bus thread into structs.

Reply:
Extracted into standalone issue #42.